packages/attune/src/apt/packages_index.rs

```diff
@@ -0,0 +1,227 @@
+                debian_repository_index_packages.md5sum,
+                debian_repository_index_packages.sha1sum,
+                debian_repository_index_packages.sha256sum
+            FROM
+                debian_repository
+                JOIN debian_repository_release ON debian_repository_release.repository_id = debian_repository.id
+                JOIN debian_repository_component ON debian_repository_component.release_id = debian_repository_release.id
+                JOIN debian_repository_index_packages ON debian_repository_index_packages.component_id = debian_repository_component.id
+            WHERE
+                debian_repository.tenant_id = $1
+                AND debian_repository.name = $2
+                AND debian_repository_release.distribution = $3
+            "#,
+            tenant_id.0,
+            repository,
+            release,
+        )
+        .fetch_all(&mut **tx)
+        .await
+        .unwrap()
+    }
+}
+
+#[derive(Clone, Debug, FromRow)]
+pub struct PackagesIndex {
+    #[sqlx(flatten)]
+    pub meta: PackagesIndexMeta,
+    pub contents: String,
+    packages: Vec<PublishedPackage>,
+}
+
+impl PackagesIndex {
+    pub fn from_packages(
+        component: &str,
+        architecture: &str,
+        packages: Vec<PublishedPackage>,
+    ) -> Self {
+        let rendered = Self::render(packages.iter());
+        Self {
+            meta: PackagesIndexMeta {
+                component: component.to_string(),
+                architecture: architecture.to_string(),
+                size: rendered.len() as i64,
+                md5sum: hex::encode(Md5::digest(&rendered)),
+                sha1sum: hex::encode(Sha1::digest(&rendered)),
+                sha256sum: hex::encode(Sha256::digest(&rendered)),
+            },
+            packages,
+            contents: rendered,
+        }
+    }
+
+    fn render<'a>(packages: impl Iterator<Item = &'a PublishedPackage>) -> String {
+        // TODO: A better way to ensure index reproducibility is to sort the
+        // packages before we render them.
+        let mut index = packages
+            .map(|published| {
+                let pkg = &published.package;
+                let fields = pkg
+                    .paragraph
+                    .as_object()
+                    .unwrap()
+                    .into_iter()
+                    .map(|(k, v)| format!("{}: {}", k, v.as_str().unwrap()))
+                    .chain(vec![
+                        format!("Filename: {}", published.filename),
+                        format!("Size: {}", pkg.size.to_string()),
+                        format!("MD5sum: {}", pkg.md5sum),
+                        format!("SHA1: {}", pkg.sha1sum),
+                        format!("SHA256: {}", pkg.sha256sum),
+                    ])
+                    .collect::<Vec<String>>();
+                fields.join("\n")
+            })
+            .collect::<Vec<String>>()
+            .join("\n\n");
```

Comment:
[fyi] since you're collecting into a vec, you can sort very cheaply using `itertools`: it won't do any extra allocations.